### ISSUE TITLE
Batched tracking broken ultralytics>=8.1.27

### ISSUE THREAD

**Issue by charlielito** [ultralytics/ultralytics#22483]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

When using batch_size>1 with `.track(persist=True)`  after version 8.1.27 (included) the returned results only include the tracking ID for the first image in the batch. For the rest, the tracking ID is just None, no tracking at all. Previous versions of the one mentioned work just fine :/

### Environment

```
Ultralytics YOLOv8.1.27 🚀 Python-3.10.12 torch-1.8.2 CUDA:0 (NVIDIA GeForce RTX 3070 Laptop GPU, 7867MiB)
Setup complete ✅ (8 CPUs, 15.3 GB RAM, 763.7/777.4 GB disk)

OS                  Linux-6.8.0-76060800daily20240311-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 15.31 GB
CPU                 11th Gen Intel Core(TM) i7-11370H 3.30GHz
CUDA                11.7

matplotlib          ✅ 3.4.2>=3.3.0
opencv-python       ✅ 4.6.0.66>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.5>=2.23.0
scipy               ✅ 1.15.3>=1.4.1
torch               ✅ 1.8.2>=1.8.0
torchvision         ✅ 0.9.2>=0.9.0
tqdm                ✅ 4.64.1>=4.64.0
psutil              ✅ 5.8.0
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 1.5.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
```

### Minimal Reproducible Example

```python
import cv2
from ultralytics import YOLO

model = YOLO("yolov8m-seg.pt")
video_paths = ["video2.mp4", "video1.mp4"]
caps = [cv2.VideoCapture(video_path) for video_path in video_paths]

while True:
    frames = []
    for cap in caps:
        ret, frame = cap.read()
        if not ret:
            print("Error: Could not read frame")
            break
        frames.append(frame)
    if len(frames) != len(caps):
        break

    results = model.track(frames,show=False, verbose=False,persist=True,tracker="bytetrack.yaml")
    for idx, result in enumerate(results):
        print("Result for image: ", idx)
        print(result.boxes.conf)
        print(result.boxes.id, "\n")
        drawn = result.plot()
        drawn = cv2.resize(drawn, dsize=None,fx=0.5,fy=0.5)
        cv2.imshow(str(idx), drawn)

    # Exit loop if the 'q' key is pressed
    if cv2.waitKey(100) & 0xFF == ord('q'):
        break

# Release the capture and close any open windows
cap.release()
cv2.destroyAllWindows()
```

Attached are the videos. You'll see visually or in the log that when using the newer version in the second video the tracking is not working. But with older versions the tracking works on both videos.

Can you point me to the release diff so maybe i could try to fix it? Thanks!

https://github.com/user-attachments/assets/ef598056-fe09-4354-aa0f-0cb0009423ae

https://github.com/user-attachments/assets/e99530ca-3da4-46c5-b3e4-f3dbbacc25a2

### Additional

With ultralytics 8.1.27 (no tracking right video)

<img width="1589" height="573" alt="Image" src="https://github.com/user-attachments/assets/db79e6e2-cee5-4140-bd12-a84490c31eb7" />

With ultralytics 8.1.26 good tracking both

<img width="1600" height="575" alt="Image" src="https://github.com/user-attachments/assets/bed69220-2777-4024-a404-ef6e1ed3182f" />

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @charlielito, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get support quickly—an Ultralytics engineer will review your report and assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for sharing clear details, the environment info, and a Minimal Reproducible Example (MRE) 🙏. If there are any additional logs or console warnings when running your snippet with `verbose=True`, please include them to help us pinpoint the issue. If you’re able, also confirm whether the behavior persists after upgrading to the latest `ultralytics` version using the steps below.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by charlielito** [comment]
I see this https://github.com/ultralytics/ultralytics/pull/8842
Ironically, that broke the example I shared 😅

**Comment by Y-T-G** [comment]
That version is over 1.5 years old. You can try the latest version to see if it still exists. It seems to be working fine in the latest version:

```python
In [1]: from ultralytics import YOLO, ASSETS
model
In [2]: model = YOLO("yolo11n.pt")

In [3]: import cv2

In [6]: frames = [cv2.imread(str(ASSETS / "bus.jpg")), cv2.imread(str(A
      ⋮ SSETS / "zidane.jpg"))]

In [7]: results = model.track(frames, persist=True, tracker="bytetrack.
      ⋮ yaml")

0: 640x640 4 persons, 1 bus, 37.3ms
1: 640x640 1 person, 37.3ms
Speed: 4.8ms preprocess, 37.3ms inference, 9.9ms postprocess per image at shape (1, 3, 640, 640)

In [8]: results[1].boxes.id
Out[8]: tensor([1.])

In [9]: results[0].boxes.id
Out[9]: tensor([1., 2., 3., 4., 5.])
```

**Comment by charlielito** [comment]
@Y-T-G  I clearly tested the latest version (8.3.221), and it didn't work either; that's why the description says versions *greater or equal* than 8.1.27, because I took the time to track down where it broke... 

I also shared a script and data to reproduce the issue, which seems to have been ignored. The snippet you shared works, but it has different conditions that don't reproduce the current bug, which, after looking at the code and the PR mentioned, is related to just having one tracker instance for all images in the batch. This breaks if the images in the batch have the same class, but they look very different.  Actually, in your results, the person with ID 1 appears in both images, but that's wrong; they are different people.

If you use the following image instead of `zidane` you'll you won't get any IDs for that image
<img width="1249" height="523" alt="Image" src="https://github.com/user-attachments/assets/a0d976e4-03ce-49fd-93ff-dac2b59d50a0" />

**Comment by Y-T-G** [comment]
@charlielito I see. I did test your code, but on a headless server. I just looked at the logs and both showed IDs at some points consistently, so I assumed it was working since you said the bug was "no tracking at all" for the second frame.

And I then tested with two images with the code I sent before, which showed IDs on both, which again was in contrast to "no tracking at all" bug report.

> Actually, in your results, the person with ID 1 appears in both images, but that's wrong; they are different people.

Technically speaking, this probably is intended and not a bug, i.e., batched frames are assumed to be sequence of frames from the same source. And independent trackers are only initialized when you use multiple streams, instead  of multiple frames. Otherwise, if you run batched inference with tracking on  a video source directly, which is a feature Ultralytics supports, with `batch` greater than 1, then the tracking would be broken.

```
# tracking would be broken if each frame in batch is treated independently
results = model.track("video1.mp4", batch=32)
```

You could limit the behavior to videos only, but then again, it depends on whether frames in a batch should be treated as frames from independent sources or as a sequence. There are users that do want sequential tracking with batched frames:
https://github.com/ultralytics/ultralytics/issues/19288

**Comment by charlielito** [comment]
@Y-T-G makes sense! 
It would be nice then to give the user the option to specify whether the frames in a batch should be treated as frames from independent sources or not. WDYT, and how could this be included? I know there is a `stream` argument in the method, but it doesn't seem to work with `.track`  
Key lines are https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/track.py#L66 and https://github.com/ultralytics/ultralytics/blob/main/ultralytics/trackers/track.py#L86 but in the example script I shared the ` predictor.dataset.mode` is set to `image` even if called with `stream=True` (maybe i don't understand well what is stream referred to)

**Comment by Y-T-G** [comment]
Yes, it triggers for streams. But it refers to `file.streams` input, which another type of inference source that Ultralytics accepts, which usually contains a list of RTSP sources.

There's a way to trigger the stream logic by manually changing the mode to `stream` using a callback at runtime:

```python
def to_stream(predictor):
   predictor.dataset.mode = "stream"

model.add_callback("on_predict_start", to_stream)
```

This should trigger the independent tracker logic and each of the frames in a batch would have a separate tracker.

**Comment by charlielito** [comment]
Nice! Amazing, that works for me!

**Comment by glenn-jocher** [comment]
Great to hear it works. If you’d like independent trackers without a callback, you can also treat each input as a stream so a separate tracker is instantiated per source; for example:

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
results = model.track(source=["video1.mp4", "video2.mp4"], stream=True, tracker="bytetrack.yaml", persist=True)
for r in results:
    pass
```

Details on how tracking is managed are in the Track mode guide and the track() API/callbacks: see the [Multi‑Object Tracking docs](https://docs.ultralytics.com/modes/track/), the [track() API reference](https://docs.ultralytics.com/reference/engine/model/#track), and the [tracking callbacks reference](https://docs.ultralytics.com/reference/trackers/track/). If you want an explicit flag to treat batched arrays as independent sources, feel free to open a small feature request for visibility.
